crates/sp-domains/src/fraud_proof.rs

```diff
@@ -181,6 +181,86 @@ pub enum VerificationError {
     Oneshot(String),
 }
 
+// TODO: Define rest of the fraud proof fields
+#[derive(Debug, Decode, Encode, TypeInfo, PartialEq, Eq, Clone)]
+pub struct MissingInvalidBundleEntryFraudProof {
+    domain_id: DomainId,
+    bundle_index: u32,
+}
+
+impl MissingInvalidBundleEntryFraudProof {
+    pub fn new(domain_id: DomainId, bundle_index: u32) -> Self {
+        Self {
+            domain_id,
+            bundle_index,
+        }
+    }
+}
+
+// TODO: Define rest of the fraud proof fields
+#[derive(Debug, Decode, Encode, TypeInfo, PartialEq, Eq, Clone)]
+pub struct AdditionalInvalidBundleEntryFraudProof {
+    domain_id: DomainId,
+    bundle_index: u32,
+}
+
+impl AdditionalInvalidBundleEntryFraudProof {
+    pub fn new(domain_id: DomainId, bundle_index: u32) -> Self {
+        Self {
+            domain_id,
+            bundle_index,
+        }
+    }
+}
+
+// TODO: Define rest of the fraud proof fields
+#[derive(Debug, Decode, Encode, TypeInfo, PartialEq, Eq, Clone)]
+pub struct MisclassifiedInvalidBundleFraudProof {
+    domain_id: DomainId,
+    bundle_index: u32,
+}
+
+impl MisclassifiedInvalidBundleFraudProof {
+    pub fn new(domain_id: DomainId, bundle_index: u32) -> Self {
+        Self {
+            domain_id,
+            bundle_index,
+        }
+    }
+}
+
+// TODO: Define rest of the fraud proof fields
+#[derive(Debug, Decode, Encode, TypeInfo, PartialEq, Eq, Clone)]
+pub struct IncorrectSortitionFraudProof {
+    domain_id: DomainId,
+}
+
+impl IncorrectSortitionFraudProof {
+    pub fn new(domain_id: DomainId) -> Self {
+        Self { domain_id }
+    }
+}
+
+/// Fraud proof indicating that `invalid_bundles` field of the receipt is incorrect
+#[derive(Debug, Decode, Encode, TypeInfo, PartialEq, Eq, Clone)]
+pub enum InvalidBundlesFraudProof {
+    MissingInvalidBundleEntry(MissingInvalidBundleEntryFraudProof),
+    AdditionalInvalidBundleEntry(AdditionalInvalidBundleEntryFraudProof),
+    MisclassifiedInvalidBundleEntry(MisclassifiedInvalidBundleFraudProof),
```

Comment:
IMO, for all these kinds of invalid bundle fraud proof, the only difference is in some fraud proof we need to prove **the bundle is indeed an invalid bundle of type `invalid_bundle_type`** and in the other case we need to prove **the bundle is not an invalid bundle of type `invalid_bundle_type`**, thus I think the following is enough:

```rust
pub strict InvalidBundlesFraudProof {
    //If `direction` is `true` means `proof` is proving the bundle is indeed an invalid bundle of type `invalid_bundle_type`, `false` means the opposite
    direction: bool
    // The concrete fraud proof of type `invalid_bundle_type`
    proof: FraudProof
}
```

Reply:
Problem with this is, actual fraud proof can have different fields depending upon which direction it is supposed to be. 

So, we need to add enum here regardless.

Reply:
Okay, in that sense, perhaps:
```rust
pub enum BundlesFraudProof {
    Valid(ValidBundlesFraudProof),
    Invalid(InvalidBundlesFraudProof),
}
```
we should try to avoid using `Missing` or `Additional` because they imply the fraud proof is targeting the whole list instead of the first mismatch.